http-gateways/REDIRECTS_FILE.md

```diff
@@ -0,0 +1,190 @@
+
+Developers can enable URL redirects or rewrites by adding redirect rules to a file named `_redirects` stored underneath the root CID of their web site.
+
+This can be used, for example, to enable URL rewriting for hosting a single-page application, to redirect invalid URLs to a pretty 404 page, or to avoid  [link rot](https://en.wikipedia.org/wiki/Link_rot) when moving to IPFS-based website hosting.
+
+# Table of Contents
+
+- [File Name and Location](#file-name-and-location)
+- [File Format](#file-format)
+  - [From](#from)
+  - [To](#to)
+  - [Status](#status)
+  - [Placeholders](#placeholders)
+  - [Splat](#splat)
+- [Evaluation](#evaluation)
+  - [Subdomain or DNSLink Gateways](#subdomain-or-dnslink-gateways)
+  - [Order](#order)
+  - [No Forced Redirects](#no-forced-redirects)
+- [Error Handling](#error-handling)
+- [Security](#security)
+- [Appendix: notes for implementors](#appendix-notes-for-implementors)
+  - [Test fixtures](#test-fixtures)
+
+# File Name and Location
+
+The Redirects File MUST be named `_redirects` and stored underneath the root CID of the web site.
+
+# File Format
+
+The Redirects File MUST be a text file containing one or more lines with the following format (brackets indication optionality).
+
+```
+from to [status]
+```
+
+## From
+
+The path to redirect from.
+
+## To
+
+The URL or path to redirect to.
+
+## Status
+
+An optional integer specifying the HTTP status code to return from the request.  Supported values are:
+
+- `200` - OK
+  - Redirect will be treated as a rewrite, returning OK without changing the URL in the browser.
+- `301` - Permanent Redirect (default)
+- `302` - Found (commonly used for Temporary Redirect)
+- `303` - See Other (replacing PUT and POST with GET)
+- `307` - Temporary Redirect (explicitly preserving body and HTTP method of original request)
+- `308` - Permanent Redirect (explicitly preserving body and HTTP method of original request)
+- `404` - Not Found
+  - Useful for redirecting invalid URLs to a pretty 404 page.
+- `410` - Gone
+- `451` - Unavailable For Legal Reasons  
+
+## Placeholders
+
+Placeholders are named variables that can be used to match path segments in the `from` path and inject them into the `to` path.
+
+For example:
+
+```
+/posts/:month/:day/:year/:slug  /articles/:year/:month/:day/:slug
+```
+
+This rule will redirect a URL like `/posts/06/15/2022/hello-world` to `/articles/2022/06/15/hello-world`.
+
+### Splat
+
+If a `from` path ends with an asterisk (i.e. `*`), the remainder of the `from` path is slurped up into the special `:splat` placeholder, which can then be injected into the `to` path.
+
+For example:
+
+```
+/posts/* /articles/:splat
+```
+
+This rule will redirect a URL like `/posts/2022/06/15/hello-world` to `/articles/2022/06/15/hello-world`.
+
+Splat logic MUST only apply to a single trailing asterisk, as this is a greedy match, consuming the remainder of the path.
```

Comment:
With a rule like `/posts/:month/:day/:year/:slug /articles/:year/:month/:day/:slug`

If a gateway receives a request `gateway.ipfs/posts/september`, should the gateway attempt to redirect to `/articles/NULL/september` ? What is the process for handling this?

Reply:
we should call out how we handle partial paths without an explicit `*` in the rule.

Reply:
> With a rule like `/posts/:month/:day/:year/:slug /articles/:year/:month/:day/:slug`
> 
> If a gateway receives a request `gateway.ipfs/posts/september`, should the gateway attempt to redirect to `/articles/NULL/september` ? What is the process for handling this?

This will not match at all since the path is missing parts of the specified `from` path.  You would have to explicitly add additional rules to handle different paths.

Reply:
> we should call out how we handle partial paths without an explicit `*` in the rule.

If I understand you correctly, there's no need to document this since there's no such thing as a partial path / partial match.  The entire `from` path must match for the rule to fire, just like the prior art we're mimicking.